File: Bibliotek/Connection/BibConnectionEndpoint.m
//
//  BibConnectionEndpoint.m
//  Bibliotek
//
//  Created by Steve Brunwasser on 8/7/18.
//  Copyright © 2018 Steve Brunwasser. All rights reserved.
//

#import "BibConnectionEndpoint.h"

NSString *const kDefaultHost = @"localhost";
NSInteger const kDefaultPort = 210;
NSString *const kDefaultDatabase = @"Default";

@implementation BibConnectionEndpoint {
@protected
    NSString *_host;
    NSInteger _port;
    NSString *_database;
    NSString *_name;
    NSString *_catalog;
}

@synthesize host = _host;
@synthesize port = _port;
@synthesize database = _database;
@synthesize name = _name;
@synthesize catalog = _catalog;

- (instancetype)init {
    return [self initWithHost:kDefaultHost port:kDefaultPort database:kDefaultDatabase];
}

- (instancetype)initWithHost:(NSString *)host {
    return [self initWithHost:host port:kDefaultPort database:kDefaultDatabase];
}

- (instancetype)initWithHost:(NSString *)host port:(NSInteger)port {
    return [self initWithHost:host port:port database:kDefaultDatabase];
}

- (instancetype)initWithHost:(NSString *)host database:(NSString *)database {
    return [self initWithHost:host port:kDefaultPort database:database];
}

- (instancetype)initWithHost:(NSString *)host port:(NSInteger)port database:(NSString *)database {
    if (self = [super init]) {
        _host = [host copy];
        _port = port;
        _database = [database copy];
    }
    return self;
}

- (instancetype)initWithEndpoint:(BibConnectionEndpoint *)endpoint {
    return [self initWithHost:[endpoint host] port:[endpoint port] database:[endpoint database]];
}

- (id)copyWithZone:(NSZone *)zone {
    return [[BibConnectionEndpoint allocWithZone:zone] initWithEndpoint:self];
}

- (id)mutableCopyWithZone:(NSZone *)zone {
    return [[BibMutableConnectionEndpoint allocWithZone:zone] initWithEndpoint:self];
}

+ (BOOL)supportsSecureCoding { return YES; }

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super init]) {
        _host = [aDecoder decodeObjectOfClass:[NSString self] forKey:@"host"];
        _port = [aDecoder decodeIntegerForKey:@"port"];
        _database = [aDecoder decodeObjectOfClass:[NSString self] forKey:@"database"];
        _name = [aDecoder decodeObjectOfClass:[NSString self] forKey:@"name"];
        _catalog = [aDecoder decodeObjectOfClass:[NSString self] forKey:@"catalog"];
    }
    return self;
}

- (void)encodeWithCoder:(NSCoder *)aCoder {
    [aCoder encodeObject:_host forKey:@"host"];
    [aCoder encodeInteger:_port forKey:@"port"];
    [aCoder encodeObject:_database forKey:@"database"];
    [aCoder encodeObject:_name forKey:@"name"];
    [aCoder encodeObject:_catalog forKey:@"catalog"];
}

- (NSString *)description {
    if (_name) {
        return [NSString stringWithFormat:@"%@ <%@:%ld/%@>", _name, _host, (long)_port, _database];
    }
    return [NSString stringWithFormat:@"%@:%ld/%@", _host, (long)_port, _database];
}

@end

@implementation BibMutableConnectionEndpoint

@dynamic host;
- (void)setHost:(NSString *)host {
    NSString *const key = NSStringFromSelector(@selector(host));
    _host = [host copy];
}

@dynamic port;
- (void)setPort:(NSInteger)port {
    NSString *const key = NSStringFromSelector(@selector(port));
    _port = port;
}

@dynamic database;
- (void)setDatabase:(NSString *)database {
    NSString *const key = NSStringFromSelector(@selector(database));
    _database = [database copy];
}

@dynamic name;
- (void)setName:(NSString *)name {
    NSString *const key = NSStringFromSelector(@selector(name));
    _name = [name copy];
}

@dynamic catalog;
- (void)setCatalog:(NSString *)catalog {
    NSString *const key = NSStringFromSelector(@selector(catalog));
    _catalog = [catalog copy];
}

@end

File: Bibliotek/Connection/BibConnectionEndpoint.swift
//
//  BibConnectionEndpoint.swift
//  Bibliotek
//
//  Created by Steve Brunwasser on 8/7/18.
//  Copyright © 2018 Steve Brunwasser. All rights reserved.
//

import Foundation

extension BibConnectionEndpoint: CustomPlaygroundDisplayConvertible {
    public var playgroundDescription: Any {
        return description
    }
}

extension BibConnectionEndpoint: Codable {
    enum CodingKeys: String, CodingKey {
        case host, port, database, name, catalog
    }

    public required convenience init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let host = try container.decode(String.self, forKey: .host)
        let port = try container.decode(Int.self, forKey: .port)
        let database = try container.decode(String.self, forKey: .database)
        let endpoint = BibMutableConnectionEndpoint(host: host, port: port, database: database)
        endpoint.name = try container.decode(String?.self, forKey: .name)
        endpoint.catalog = try container.decode(String?.self, forKey: .catalog)
        self.init(endpoint)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(host, forKey: .host)
        try container.encode(port, forKey: .port)
        try container.encode(database, forKey: .database)
        try container.encode(name, forKey: .name)
        try container.encode(catalog, forKey: .catalog)
    }
}

// MARK: - Swift Native

extension Connection {
    public struct Endpoint {
        private var endpoint: BibMutableConnectionEndpoint

        internal init(endpoint: BibConnectionEndpoint) {
            self.endpoint = endpoint.mutableCopy() as! BibMutableConnectionEndpoint
        }

        public init(host: String = "localhost", port: Int = 210, database: String = "Default") {
            endpoint = BibMutableConnectionEndpoint(host: host, port: port, database: database)
        }
    }
}

extension Connection.Endpoint {
    public var host: String {
        get { return endpoint.host }
        set {
            if !isKnownUniquelyReferenced(&endpoint) {
                endpoint = endpoint.mutableCopy() as! BibMutableConnectionEndpoint
            }
            endpoint.host = newValue
        }
    }

    public var port: Int {
        get { return endpoint.port }
        set {
            if !isKnownUniquelyReferenced(&endpoint) {
                endpoint = endpoint.mutableCopy() as! BibMutableConnectionEndpoint
            }
            endpoint.port = newValue
        }
    }

    public var database: String {
        get { return endpoint.database }
        set {
            if !isKnownUniquelyReferenced(&endpoint) {
                endpoint = endpoint.mutableCopy() as! BibMutableConnectionEndpoint
            }
            endpoint.database = newValue
        }
    }

    public var name: String? {
        get { return endpoint.name }
        set {
            if !isKnownUniquelyReferenced(&endpoint) {
                endpoint = endpoint.mutableCopy() as! BibMutableConnectionEndpoint
            }
            endpoint.name = newValue
        }
    }

    public var catalog: String? {
        get { return endpoint.catalog }
        set {
            if !isKnownUniquelyReferenced(&endpoint) {
                endpoint = endpoint.mutableCopy() as! BibMutableConnectionEndpoint
            }
            endpoint.catalog = newValue
        }
    }
}

extension Connection.Endpoint: CustomPlaygroundDisplayConvertible {
    public var playgroundDescription: Any {
        return endpoint.playgroundDescription
    }
}

extension Connection.Endpoint: CustomStringConvertible {
    public var description: String {
        return endpoint.description
    }
}

extension Connection.Endpoint: Codable {
    public init(from decoder: Decoder) throws {
        endpoint = try BibMutableConnectionEndpoint(from: decoder)
    }

    public func encode(to encoder: Encoder) throws {
        try endpoint.encode(to: encoder)
    }
}

// MARK: - Objective-C Bridging

extension Connection.Endpoint: _ObjectiveCBridgeable {
    public func _bridgeToObjectiveC() -> BibConnectionEndpoint {
        return endpoint.copy() as! BibConnectionEndpoint
    }

    public static func _forceBridgeFromObjectiveC(_ source: BibConnectionEndpoint, result: inout Connection.Endpoint?) {
        result = Connection.Endpoint(endpoint: source)
    }

    public static func _conditionallyBridgeFromObjectiveC(_ source: BibConnectionEndpoint, result: inout Connection.Endpoint?) -> Bool {
        result = Connection.Endpoint(endpoint: source)
        return true
    }

    public static func _unconditionallyBridgeFromObjectiveC(_ source: BibConnectionEndpoint?) -> Connection.Endpoint {
        return source.map(Connection.Endpoint.init(endpoint:)) ?? Connection.Endpoint()
    }
}
